### PR TITLE
ISPN-11075 Server security usability and docs

### PR DESCRIPTION
**Pull request by tristantarrant**
https://issues.redhat.com/browse/ISPN-11073
ISPN-11073 Add a TokenCallbackHandler to simplify OAUTHBEARER configuration
https://issues.redhat.com/browse/ISPN-11074
ISPN-11074 LDAP Realm connection configuration
https://issues.redhat.com/browse/ISPN-11075
ISPN-11075 Update the Server security documentation
https://issues.redhat.com/browse/ISPN-11089
ISPN-11089 LDAP Attribute reference

### CODE REVIEW

File: client/hotrod-client/src/main/java/org/infinispan/client/hotrod/configuration/AuthenticationConfigurationBuilder.java [hunk 1]
```diff
@@ -173,20 +175,31 @@ public AuthenticationConfigurationBuilder password(char[] password) {
 
    /**
     * Specifies the realm to be used for authentication. Username and password also need to be supplied. If none
-    * is specified, this defaults to 'ApplicationRealm'.
+    * is specified, this defaults to 'default'.
```
Comment (karesti) [inline client/hotrod-client/src/main/java/org/infinispan/client/hotrod/configuration/AuthenticationConfigurationBuilder.java]: can we call the Constant from this doc ?

File: client/hotrod-client/src/main/java/org/infinispan/client/hotrod/configuration/AuthenticationConfigurationBuilder.java [hunk 2]
```diff
@@ -173,20 +175,31 @@ public AuthenticationConfigurationBuilder password(char[] password) {
...
 
+   public AuthenticationConfigurationBuilder token(String token) {
+      this.token = token;
+      return enable();
+   }
+
    @Override
    public AuthenticationConfiguration create() {
-      String mech = saslMechanism == null ? "DIGEST-MD5" : saslMechanism;
+      String mech = saslMechanism == null ? "SCRAM-SHA-512" : saslMechanism;
```
Comment (karesti) [inline client/hotrod-client/src/main/java/org/infinispan/client/hotrod/configuration/AuthenticationConfigurationBuilder.java]: constant string in a single place?

File: client/hotrod-client/src/main/java/org/infinispan/client/hotrod/configuration/AuthenticationConfigurationBuilder.java [hunk 3]
```diff
@@ -173,20 +175,31 @@ public AuthenticationConfigurationBuilder password(char[] password) {
...
       CallbackHandler cbh = callbackHandler;
       if (cbh == null) {
-         if (username != null) {
+         if ("OAUTHBEARER".equals(mech)) {
```
Comment (karesti) [inline client/hotrod-client/src/main/java/org/infinispan/client/hotrod/configuration/AuthenticationConfigurationBuilder.java]: idem constant string in single place ?

File: client/hotrod-client/src/main/java/org/infinispan/client/hotrod/configuration/AuthenticationConfigurationBuilder.java (hunk 2 above)
Comment (tristantarrant) [reply client/hotrod-client/src/main/java/org/infinispan/client/hotrod/configuration/AuthenticationConfigurationBuilder.java]: Why constify this ?

Comment (tristantarrant) [reply client/hotrod-client/src/main/java/org/infinispan/client/hotrod/configuration/AuthenticationConfigurationBuilder.java]: These are SASL mechanism names which don't change. Constifying them makes no sense and would make the examples unreadable.

**Comment by tristantarrant** [comment]
@oraNod I've added documentation on connecting to Active Directory as well as OpenLDAP/RHDS/ADS.

File: documentation/src/main/asciidoc/topics/server_protocols_hotrod.adoc [hunk 4]
```diff
@@ -45,18 +48,46 @@ clientBuilder
...
+clientBuilder
+    .addServer()
+        .host("127.0.0.1")
+        .port(11222)
+    .security()
+        .authentication()
+            .saslMechanism("DIGEST-MD5")
+            .username("myuser")
+            .password("qwer1234!")
+            .realm("default");
```
Comment (galderz) [inline documentation/src/main/asciidoc/topics/server_protocols_hotrod.adoc]: Why is `realm` spelled out in this example, even if it contains the default value, but not server name? The same happens with other examples throughout the documentation.

Comment (tristantarrant) [reply documentation/src/main/asciidoc/topics/server_protocols_hotrod.adoc]: I have removed all examples which use the defaults. I've also externalized all of the code examples to their own individual files.

**Comment by tristantarrant** [comment]
This should be good to go

File: client/hotrod-client/src/main/java/org/infinispan/client/hotrod/configuration/AuthenticationConfigurationBuilder.java (hunk 2 above)
Comment (tristantarrant) [reply client/hotrod-client/src/main/java/org/infinispan/client/hotrod/configuration/AuthenticationConfigurationBuilder.java]: Apologies, I thought you were referencing examples. Fixed.

File: client/hotrod-client/src/main/java/org/infinispan/client/hotrod/configuration/AuthenticationConfigurationBuilder.java (hunk 3 above)
Comment (tristantarrant) [reply client/hotrod-client/src/main/java/org/infinispan/client/hotrod/configuration/AuthenticationConfigurationBuilder.java]: Done

File: client/hotrod-client/src/main/java/org/infinispan/client/hotrod/configuration/AuthenticationConfigurationBuilder.java (hunk 1 above)
Comment (tristantarrant) [reply client/hotrod-client/src/main/java/org/infinispan/client/hotrod/configuration/AuthenticationConfigurationBuilder.java]: Done

File: documentation/src/main/asciidoc/topics/config_examples/server_ldap_realm_ad.xml [hunk 5]
```diff
@@ -0,0 +1,17 @@
+<security xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
```
Comment (pruivo) [inline documentation/src/main/asciidoc/topics/config_examples/server_ldap_realm_ad.xml]: where is this example used?

File: server/runtime/src/main/java/org/infinispan/server/configuration/Attribute.java [hunk 6]
```diff
@@ -79,7 +83,7 @@
    WORKER_THREADS,
    WRAP_GSS_CREDENTIAL,
    WRITABLE,
-   ;
+   SEARCH_RECURSIVE, ROLE_RECURSION, ROLE_RECURSION_NAME, EXTRACT_RDN, REFERENCE;
```
[the remark this answers is not in the document]
Comment (tristantarrant) [reply server/runtime/src/main/java/org/infinispan/server/configuration/Attribute.java]: Sorted !

File: server/runtime/src/main/java/org/infinispan/server/configuration/security/LdapAttributeConfigurationBuilder.java [hunk 7]
```diff
@@ -65,4 +100,6 @@ public LdapAttributeConfigurationBuilder read(LdapAttributeConfiguration templat
       this.attributes.read(template.attributes());
       return this;
    }
+
```
[the remark this answers is not in the document]
Comment (tristantarrant) [reply server/runtime/src/main/java/org/infinispan/server/configuration/security/LdapAttributeConfigurationBuilder.java]: Removed

File: documentation/src/main/asciidoc/topics/security_server.adoc [hunk 8]
```diff
@@ -11,45 +11,295 @@ include::config_examples/cache_authorization.xml[]
...
+
+.Creating certificates with keytool
+[source,options="nowrap",subs=attributes+]
+----
+include::cmd_examples/keytool.adoc[]
+----
+
+Place the `server.p12` file in the server configuration directory.
+
+NOTE: To configure the client In order to connect to the server using the Hot Rod protocol, the client needs a trust store containing the public key
```
[the remark this answers is not in the document]
Comment (tristantarrant) [reply documentation/src/main/asciidoc/topics/security_server.adoc]: Fixed

File: documentation/src/main/asciidoc/topics/security_server.adoc [hunk 9]
```diff
@@ -11,45 +11,295 @@ include::config_examples/cache_authorization.xml[]
...
+You can require that clients also present a certificate by adding `require-ssl-client-auth="true"` to the `endpoint` element:
+
+.Requiring client certificates
+[source,xml,options="nowrap",subs=attributes+]
+----
+include::config_examples/server_endpoint_require_client_cert.xml[]
+----
+
+It is also possible to generate development certificates on server startup. In order to do this, just specify `generate-self-signed-certificate-host`
+in the keystore element as shown below:
```
[the remark this answers is not in the document]
Comment (tristantarrant) [reply documentation/src/main/asciidoc/topics/security_server.adoc]: Added

File: documentation/src/main/asciidoc/topics/cmd_examples/keytab_nix_mit.adoc [hunk 10]
```diff
@@ -0,0 +1,5 @@
+$ ktutil
+ktutil:  addent -password -p datagrid@DOMAIN.ORG -k 1 -e aes256-cts
```
[the remark this answers is not in the document]
Comment (tristantarrant) [reply documentation/src/main/asciidoc/topics/cmd_examples/keytab_nix_mit.adoc]: Renamed

File: documentation/src/main/asciidoc/topics/cmd_examples/keytab_win.adoc [hunk 11]
```diff
@@ -0,0 +1,2 @@
+$ ktpass -princ HTTP/datagrid@DOMAIN.ORG -pass * -mapuser DOMAIN\USER_NAME
```
[the remark this answers is not in the document]
Comment (tristantarrant) [reply documentation/src/main/asciidoc/topics/cmd_examples/keytab_win.adoc]: Renamed

File: documentation/src/main/asciidoc/topics/config_examples/server_ldap_realm_ad.xml (hunk 5 above)
Comment (tristantarrant) [reply documentation/src/main/asciidoc/topics/config_examples/server_ldap_realm_ad.xml]: I had removed it by accident. Added to the security_server.adoc

File: documentation/src/main/asciidoc/topics/security_server.adoc [hunk 12]
```diff
@@ -11,45 +11,295 @@ include::config_examples/cache_authorization.xml[]
...
+the JRE.
+
+The following code snippet shows how to use the `client_truststore.p12` we generated above in your clients so that they will trust the server's certificate:
+
+[source,java]
+----
+include::code_examples/HotRodTrustStore.java[]
+----
+
+You can require that clients also present a certificate by adding `require-ssl-client-auth="true"` to the `endpoint` element:
```
[the remark this answers is not in the document]
Comment (tristantarrant) [reply documentation/src/main/asciidoc/topics/security_server.adoc]: Fixed

File: client/hotrod-client/src/main/java/org/infinispan/client/hotrod/security/TokenCallbackHandler.java [hunk 13]
```diff
@@ -0,0 +1,37 @@
...
+
+import org.wildfly.security.auth.callback.CredentialCallback;
+import org.wildfly.security.credential.BearerTokenCredential;
+
+/**
+ * @author Tristan Tarrant &lt;tristan@infinispan.org&gt;
+ * @since 10.1
+ **/
+public class TokenCallbackHandler implements CallbackHandler {
+   private String token;
```
[the remark this answers is not in the document]
Comment (tristantarrant) [reply client/hotrod-client/src/main/java/org/infinispan/client/hotrod/security/TokenCallbackHandler.java]: Good point. Fixed.

File: client/hotrod-client/src/main/java/org/infinispan/client/hotrod/configuration/AuthenticationConfigurationBuilder.java [hunk 14]
```diff
@@ -173,26 +182,37 @@ public AuthenticationConfigurationBuilder password(char[] password) {
...
+
    @Override
    public AuthenticationConfiguration create() {
-      String mech = saslMechanism == null ? "DIGEST-MD5" : saslMechanism;
+      String mech = saslMechanism == null ? DEFAULT_MECHANISM : saslMechanism;
       CallbackHandler cbh = callbackHandler;
       if (cbh == null) {
-         if (username != null) {
+         if (OAUTHBEARER_MECH.equals(mech)) {
+            if (token != null) {
```
[the remark this answers is not in the document]
Comment (tristantarrant) [reply client/hotrod-client/src/main/java/org/infinispan/client/hotrod/configuration/AuthenticationConfigurationBuilder.java]: Moved

**Comment by pruivo** [comment]
integrated! thanks @tristantarrant !

**Pull request merged**